### DISCUSSION TITLE
Changes to 4swap Swap Action for Improved Performance

### DISCUSSION THREAD

**Discussion by lyricat** [pandodao/developers.pando.im#3] [Announcements]
For the reasons of performance, 4swap will make the following changes to the swap action:

In the event that route_hash is empty in the [swap action](https://developers.pando.im/references/4swap/action.html#swap), 4swap will execute the swap in the pool of the from and to tokens. Please note that in this scenario, 4swap will not choose the best route.

Furthermore, if there is no pool available for the from and to tokens, the swap will fail and a refund will be issued.

Refer to: https://developers.pando.im/guide/changelog.html#_1-0-0

**Comment by coolit** [comment]
Hello there. My project was written in Rust, but I used 4swap-sdk-go to generate memo. What should I do now to adapt to this change? I cannot find the documentation for route_hash. Thanks.

**Reply by lyricat** [reply]
Please follow the [guide](https://developers.pando.im/guide/invoke-actions.html#generate-actions-by-sdk), call `fswap.Route()` and read ` preOrder.Routes` to get the calculated route.

If you prefer to implement the algorithm in Rust, [this link](https://github.com/fox-one/4swap-sdk-go/blob/e62757b2c4966d3ebac7eb40dcad7d1926c7f9e3/route.go#L169) would be helpful.
